### PR TITLE
added contracts entry point

### PR DESCRIPTION
**Pull request by ckatsube**
## Motivation and Context

See #778 for an introduction to the motivation and context.
778 was closed due to the approach where decompiling the file into its AST changed the line numbers, resulting in unhelpful error tracebacks.

## Your Changes

**Description**:

Added `__main__.py` for the contracts module which takes in the path to Python script to be executed as main, additional module names to be checked, and whether the main module should be checked.

Used a pseudo module to replace the main entry point for contract checking the main module

Alternatives:
- could have set checking main to be permanently false and added the pseudo module to the list of extra imports and used the cli flag to determine whether or not to add the pseudo module to the list or not
- ~~can just have one layer of execs and manually pass in the context dictionaries into the before and after execs.~~

**Type of change** (select all that apply):

- [x] New feature (non-breaking change which adds functionality)

## Testing

Confirmed test suite still runs.
Used two modules (not part of PR) to test the successful assertion error on contract violation for both a main and extra module.

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the CI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls.

### CODE REVIEW

**Comment by ckatsube** [comment]
Changelog:
- Added CLI sections on the `contracts/index.md` documentation page.
- Added proper exit codes
- Refactored out `astroid` with vanilla Python processing
- Updated the name for the flag which disables contract checking for the main module
- Refactored imports
- Updated docstring to be more accurate

File: python_ta/contracts/__main__.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import traceback
+import types
+from typing import List, TextIO, Tuple
+
+import astroid
+import click
+
+
+@click.command()
+@click.argument("main", type=click.File(mode="r"))
```
[the remark this answers is not in the document]
Comment (ckatsube) [reply python_ta/contracts/__main__.py]: Click arguments don't have a `help` parameter. It seems like `click` suggests including documentation for arguments inside the docstring, as the docstring also appears when using the `--help` flag.
I think I can convert the `EXTRA_MOD_NAMES` to be a flag instead (similar to Pytest's `--ignore=path` flag), but I can't see `FILE` not being an argument.

Source: https://click.palletsprojects.com/en/8.0.x/documentation/#documenting-arguments

File: docs/contracts/index.md [hunk 2]
```diff
@@ -26,6 +26,11 @@ if __name__ == '__main__':
     divide(10, 0)  # Preconditions on y violated
 ```
 
+```{note}
```
Comment (david-yz-liu) [inline docs/contracts/index.md:29]: You can remove this note, I don't want to distract from the main flow of the page here. It's good to show an example, but this should be done in the CLI section below.

File: docs/contracts/index.md [hunk 3]
```diff
@@ -67,6 +72,20 @@ Finally, you can set the `DEBUG_CONTRACTS` constant to `True` to enable debuggin
 .. autodata:: python_ta.contracts.DEBUG_CONTRACTS
 ```
 
+## CLI
```
Comment (david-yz-liu) [inline docs/contracts/index.md:75]: Don't use the acronym here, Instead, use "Command-line interface".

File: docs/contracts/index.md [hunk 4]
```diff
@@ -486,3 +505,10 @@ The above example outputs the following
 >>> Properly raised AssertionError
 >>> Did not raise AssertionError
 ```
+
+### Contracts CLI Limitations
```
Comment (david-yz-liu) [inline docs/contracts/index.md:509]: Don't make this a separate section. Instead, you can make this a "note" in the CLI section.

File: docs/contracts/index.md [hunk 5]
```diff
@@ -67,6 +72,20 @@ Finally, you can set the `DEBUG_CONTRACTS` constant to `True` to enable debuggin
 .. autodata:: python_ta.contracts.DEBUG_CONTRACTS
 ```
 
+## CLI
+
+The `python_ta.contracts` CLI can execute a file as `__main__` with contracts enabled. This option
```
Comment (david-yz-liu) [inline docs/contracts/index.md:77]: The second sentence ("This option...") is redundant with the "limitations" section below. Remove this sentence.

File: python_ta/contracts/__main__.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    FILE the Python script as if you were to just run `python FILE`
+    EXTRA_MOD_NAMES are the module names to also have contracts checked.
+        These are directly passed in, so the format is the module names as if you were importing
+    """
+    contents = file.read()
+    lines = contents.splitlines()
+
+    main_lineno = _find_main_lineno(lines)
+
+    def _formatted_traceback() -> str:
```
Comment (david-yz-liu) [inline python_ta/contracts/__main__.py:28]: Don't make this a nested function, move it to the global scope.

File: python_ta/contracts/__main__.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+from typing import List, TextIO, Tuple
+
+import click
+
+from . import check_all_contracts
+
+
+@click.command()
+@click.argument("file", type=click.File(mode="r"))
+@click.argument("extra-mod-names", nargs=-1)
```
Comment (david-yz-liu) [inline python_ta/contracts/__main__.py:14]: Yeah, you can make this an option. You can make it a comma-separated list of module names.

File: python_ta/contracts/__main__.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+
+def _find_main_lineno(lines: list[str]) -> int:
+    for lineno, line in enumerate(lines, start=1):
+        if _has_main_check(line):
+            return lineno
+    return 0
+
+
+def _has_main_check(line: str):
```
Comment (david-yz-liu) [inline python_ta/contracts/__main__.py:97]: Add a return type here

File: python_ta/contracts/__main__.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    if line.strip() == "":
+        return False
+    keyword, *condition = line.split()
```
Comment (david-yz-liu) [inline python_ta/contracts/__main__.py:100]: To be safe, also check that the line starts at the top level (i.e., there's no indentation).

File: docs/contracts/index.md [hunk 10]
```diff
@@ -67,6 +72,20 @@ Finally, you can set the `DEBUG_CONTRACTS` constant to `True` to enable debuggin
...
+requires there to be an `if __name__ == "__main__":` so PythonTA knows when and where to
+start tracking contracts from.
+
+The syntax for this command is: `python -m python_ta.contracts [OPTIONS] FILE [EXTRA_MOD_NAMES]`
```
Comment (david-yz-liu) [inline docs/contracts/index.md:81]: You don't need to specify the options here. Instead, show just the basic usage, and then say that the user can run `python_ta.contracts --help` for information about the command-line options.

File: docs/contracts/index.md [hunk 11]
```diff
@@ -67,6 +72,20 @@ Finally, you can set the `DEBUG_CONTRACTS` constant to `True` to enable debuggin
...
+
+where `FILE` is the main script, `EXTRA_MOD_NAMES` are the names of imported modules to also be checked,
+and `--no-decorate-main`, the only available flag, is whether to decorate the main script.
+
+The utmost basic usage of this command is with `python -m python_ta.contracts FILE`.
```
Comment (david-yz-liu) [inline docs/contracts/index.md:86]: This sentence should be merged with the previous paragraph.

File: docs/contracts/index.md (hunk 10 above)
Comment (david-yz-liu) [reply docs/contracts/index.md:81]: Oh and say that the user might need to use `python3` instead of `python` depending on their system.

**Review by david-yz-liu — COMMENTED** [review]
@ckatsube Great, I left some more comments. Also please update the Changelog to add an entry for this update.

**Comment by ckatsube** [comment]
Changelog: 
- Changed extra module argument to option (had to make it a multiple option since options don't support `nargs=-1`)
- Refactored traceback formatting helper to be global instead of local
- Main if block check now includes making sure the statement is at the module level
- Consolidated some text on the contract's documentation page to be more together and specific.

File: python_ta/contracts/__main__.py (hunk 9 above)
Comment (david-yz-liu) [reply python_ta/contracts/__main__.py:99]: @ckatsube I'm not sure if you missed this comment, or had a question about it?

Comment (ckatsube) [reply python_ta/contracts/__main__.py:99]: The change I made was a few more lines down (line 103 has `line.startswith(keyword)`).

**Pull request merged**